csrc/preseg_passes/insert_reshardings.cpp

```diff
@@ -70,7 +70,19 @@ void insertReshardingSetsBefore(Fusion* fusion) {
       new_inputs.push_back(new_input);
       expr = ir_utils::replaceValInExprInputs(expr, input, new_input);
     }
-    shardAllLike(output, new_inputs);
+
+    // Consider a reshard before case:
+    // A [DIDx(i0), i1] + B [i0, i1] = C [i0, DIDx(i1)]
+    // This is decomposed into:
+    // A [DIDx(i0), i1] -> set -> new_A [i0, DIDx(i1)]
+    // B [i0, i1] -> set -> new_B [i0, DIDx(i1)]
+    // new_A [i0, DIDx(i1)] + new_B [i0, DIDx(i1)] = C [i0, DIDx(i1)]
+    // ParallelType::Serial is required here so the
+    // A is sharded as [i0, DIDx(i1)] instead of [DIDx(i0), DIDx(i1)]
+    std::unordered_set<ParallelType> parallel_types{
+        kParallelTypeDIDs.begin(), kParallelTypeDIDs.end()};
+    parallel_types.insert(ParallelType::Serial);
```

Comment:
Nit: I'm on the fence between adding Serial inside shardAllLike vs letting caller adding Serial. The former is implicit and maybe surprising. The latter is explicit but requires more typing. 

I'm totally fine with the change as is. I'm mentioning this just as a trade-off to consider for the future.

Reply:
I was surprised to see Serial in the implementation earlier, and `insert_resharding` is a good place to explain the behavior, and so far, the place we require it, so I prefer keeping serial our of `shardAllLike`.
If we have several instances where this is required, we can switch the behavior